core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -181,7 +181,13 @@ public void start() {
       advancedListener = new AdvancedPurgeListener<>(expirationManager.wired());
       preloaded = false;
       enabled = configuration.persistence().usingStores();
-      blockingScheduler = Schedulers.from(blockingExecutor);
+      blockingScheduler = Schedulers.from(task -> {
+         if (isCurrentThreadBlocking()) {
+            SecurityActions.startThread(task, "persistence-bulk-blocking-operation");
```

Comment:
This is definitely worth a comment. Does RxJava guarantee that it will request a new thread from the scheduler only once, on the thread that subscribes to the `Flowable`, or could it be on a non-blocking thread when using `continueOnCPUExecutor`?

I was thinking of a different workaround: check whether the caller thread is a blocking thread once, at the beginning of every method, and pass that information in order to avoid submitting a task to the blocking executor **and** submitting a task to the non-blocking executor (if we switched to a non-blocking thread in the middle of the bulk operation, we'd need a new blocking thread for the next blocking task).

Reply:
> This is definitely worth a comment. Does RxJava guarantee that it will request a new thread from the scheduler only once, on the thread that subscribes to the `Flowable`, or could it be on a non-blocking thread when using `continueOnCPUExecutor`?

This is standard rxjava stuff with schedulers and how reactive streams work. You only subscribe to the publisher once, thus only one blocking thread is in use. Observation is also only a single thread at a time (processing a queue) However a non blocking thread will never subscribe to a flowable as it is only on publishing of a value.

The only way subscription will happen concurrently is if you use the `parallel` method or `flatMap` with `subscribeOn` in the resulting Publishers.

> I was thinking of a different workaround: check whether the caller thread is a blocking thread once, at the beginning of every method, and pass that information in order to avoid submitting a task to the blocking executor **and** submitting a task to the non-blocking executor (if we switched to a non-blocking thread in the middle of the bulk operation, we'd need a new blocking thread for the next blocking task).

This is actually what I have in my pending work.

Also you don't "switch" between blocking threads, there would only be one. And you don't "switch" to a non blocking thread in the middle. It is only publishing of a value, so it is only operators below that. You can think of it this way that `subscribeOn` allows a different thread to process all operators until it is fully published or it runs into a `subscribeOn`. In the latter it just passes the value to a queue to be processed by the other thread. And note `observeOn` will only allow 1 thread to be processing at a time per Publisher. Everything is per publisher.

Reply:
> The only way subscription will happen concurrently is if you use the `parallel` method or `flatMap` with `subscribeOn` in the resulting Publishers.

Yeah, that's what I was thinking about. What if the caller does use `parallel`?

> This is actually what I have in my pending work.

Ok then :)